docker/docker_utils.go

```diff
@@ -1,20 +1,32 @@
 package docker
 
-import "github.com/docker/docker/reference"
+import (
+	"fmt"
 
-// parseImageName converts a string into a reference and tag value.
-func parseImageName(img string) (reference.Named, string, error) {
+	"github.com/docker/docker/reference"
+)
+
+// parseImageName converts a string into a reference.
+// It is guaranteed that reference.IsNameOnly is false for the returned value.
+func parseImageName(img string) (reference.Named, error) {
 	ref, err := reference.ParseNamed(img)
 	if err != nil {
-		return nil, "", err
+		return nil, err
 	}
 	ref = reference.WithDefaultTag(ref)
-	var tag string
-	switch x := ref.(type) {
-	case reference.Canonical:
-		tag = x.Digest().String()
-	case reference.NamedTagged:
-		tag = x.Tag()
+	if reference.IsNameOnly(ref) { // Sanity check that we are fulfulling our contract
+		return nil, fmt.Errorf("Internal inconsistency: reference.IsNameOnly for reference %s (parsed from %s)", ref.String(), img)
+	}
+	return ref, nil
+}
+
+// tagOrDigest returns a tag or digest from a reference for which !reference.IsNameOnly.
```

Comment:
do we still handle the case where a tag(or digest) is entirely missing and we set `latest` as tag? https://github.com/containers/image/blob/master/docker/docker_utils.go#L11